shoulder/generator/cxx_header_generator.py

```diff
@@ -48,350 +46,403 @@ def generate(self, objects, outpath):
             )
             raise ShoulderGeneratorException(msg)
 
-    def _generate_cxx_includes(self, outfile):
-        cxx_includes = "#include <stdint>\n"
-        cxx_includes += "#include \"{path}\"\n".format(path=config.regs_h_name)
-        outfile.write(cxx_includes)
-        outfile.write("\n")
-
-    def _generate_namespace_open(self, outfile):
-        if not config.cxx_namespace: return
-        namespace_str = str(config.cxx_namespace)
-        namespaces = namespace_str.split("::")
-        if namespaces[0]:
-            for namespace in namespaces:
-                outfile.write("namespace " + namespace + "\n{\n")
-
-    def _generate_namespace_close(self, outfile):
-        if not config.cxx_namespace: return
-        namespace_str = str(config.cxx_namespace)
-        namespaces = namespace_str.split("::")
-        for namespace in namespaces:
-            outfile.write("}\n")
-
     @license
     @include_guard
-    def _generate_objects(self, objects, outfile):
-        self._generate_cxx_includes(outfile)
-        self._generate_namespace_open(outfile)
-
+    @header_depends
+    @cxx.namespace(name=config.cxx_namespace)
+    @accessor_macros
+    def _generate(self, outfile, objects):
+        """
+        Generate for all objects passed into this generator.  Wrap all generated
+        content in a license, include guard, and namespace. Append low-level
+        accessor macros before generating register accessors
+        """
         for obj in objects:
             if(isinstance(obj, Register)):
+                if not obj.is_sysreg: return
+
                 logger.debug("Writing register: " + str(obj.name))
-                self._generate_register(obj, outfile)
+                self._generate_register(outfile, obj)
+
             else:
                 msg = "Cannot generate object of unsupported {t} type".format(
                     t = type(obj)
                 )
                 raise ShoulderGeneratorException(msg)
 
-        self._generate_namespace_close(outfile)
-
-    def _generate_register(self, reg, outfile):
-        reg_cxx_name = str(reg.name).lower()
-        reg_cxx_size = "uint64_t" if reg.size == 64 else "uint32_t"
-
-        outfile.write("\n")
-        reg_comment = "// {name} ({long_name})\n// {purpose}\n".format(
-            name = str(reg.name),
-            long_name = str(reg.long_name),
-            purpose = str(reg.purpose)
-        )
-        outfile.write(reg_comment)
-
-        reg_namespace = "namespace {name}\n{{\n".format(name=reg_cxx_name)
-        outfile.write(reg_namespace)
-
-        self._increase_indent()
+    def _generate_register(self, outfile, reg):
+        """
+        Generate accessors for a single register
+        """
+        self._generate_register_comment(outfile, reg)
```

Comment:
I like moving these long print formatting blocks into separate functions. A lot easier to read and a lot easier to edit